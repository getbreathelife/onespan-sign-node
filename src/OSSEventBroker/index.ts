export * from './OSSEventBroker';
export * from './error';
