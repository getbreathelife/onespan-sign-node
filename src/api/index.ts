import * as RequestBuilders from './requestBuilders';

export { Api } from './Api';
export { RequestBuilders };
