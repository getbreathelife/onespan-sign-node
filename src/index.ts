/**
 * A library to interact with OneSpan Sign's API.
 *
 * @example
 * ```ts
 * const oneSpanSign = new OneSpanSign({
 *   clientId: CLIENT_ID,
 *   secret: CLIENT_SECRET,
 *   type: 'OWNER',   // or 'SENDER'
 * });
 *
 * const { id: packageId } = await oneSpanSign.packages.create({...});
 *
 * await oneSpanSign.documents.create(packageId, ...)
 * ```
 *
 * @packageDocumentation
 */

export { OneSpanSign } from './OneSpanSign';
export * from './api';
export * from './resources';
export * from './types';
export * from './types/requests';
