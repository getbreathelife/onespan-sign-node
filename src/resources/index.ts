export { Resource } from './Resource';
export { DocumentResource } from './DocumentResource';
export { PackageResource } from './PackageResource';
