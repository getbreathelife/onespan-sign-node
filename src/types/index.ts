// Entity typing reference: https://community.onespan.com/sites/default/files/interactive_api_references/oss_1145.yaml
import * as Events from './events';
import * as Requests from './requests';
import * as Responses from './responses';

export * from './api';
export * from './models';
export * from './utils';
export { Requests, Responses, Events };
