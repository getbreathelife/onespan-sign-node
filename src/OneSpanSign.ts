import FormData from 'form-data';

import { Api } from './api';
import {
  CreatePackageRequestPayload,
  CreatePackageResponsePayload,
  DocumentMetadata,
  UploadDocumentRequestPayload,
} from './types';

export class OneSpanSign {
  constructor(private apiKey: string, private apiUrl: string) {}

  public async createPackage(payload: CreatePackageRequestPayload): Promise<CreatePackageResponsePayload> {
    const response = await Api.post(`${this.apiUrl}/api/packages`)
      .withAuthorizationHeader(`Basic ${this.apiKey}`)
      .withBody(payload)
      .fetch();

    return (await response.json()) as CreatePackageResponsePayload;
  }

  public async uploadDocument(
    packageId: string,
    payload: UploadDocumentRequestPayload,
    documentBody: Buffer | ReadableStream
  ): Promise<DocumentMetadata> {
    const formData = new FormData();
    formData.append('file', documentBody, { filename: payload.name });
    formData.append('payload', payload);

    const response = await Api.post(`${this.apiUrl}/api/packages/${packageId}/documents`)
      .withAuthorizationHeader(`Basic ${this.apiKey}`)
      .withBody(formData)
      .fetch();

    return (await response.json()) as DocumentMetadata;
  }
}
