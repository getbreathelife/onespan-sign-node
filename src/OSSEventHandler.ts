import { EventHandler, Events } from 'types';

/**
 * Helper class to ingest
 * {@link https://community.onespan.com/documentation/onespan-sign/guides/feature-guides/developer/setting-callback-notifications | callback notifications}
 * from OneSpan Sign and call the appropriate handler.
 *
 * @public
 */
export class OSSEventBroker {
  protected _documentSignedHandler?: EventHandler<Events.DocumentSignedEvent>;

  public get documentSignedHandler(): EventHandler<Events.DocumentSignedEvent> | undefined {
    return this._documentSignedHandler;
  }

  public set documentSignedHandler(handler) {
    this._documentSignedHandler = handler;
  }
}
