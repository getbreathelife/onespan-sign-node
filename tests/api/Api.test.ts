import { AccessTokenOwnerConfig, Api } from '../../src';
import { mockedFetch, mockFetchHappyPath } from '../setup/mockNodeFetch';

const MOCK_TOKEN_CONFIG: AccessTokenOwnerConfig = {
  type: 'OWNER',
  clientId: 'clientId',
  secret: 'secret',
};
const MOCK_API_URL = 'http://demo.com';

describe('Api', () => {
  describe('Authorization header', () => {
    let api: Api;

    beforeAll(() => {
      mockFetchHappyPath();
      api = new Api(MOCK_TOKEN_CONFIG, MOCK_API_URL);
    });

    afterAll(() => {
      mockedFetch.mockReset();
    });

    afterEach(() => {
      mockedFetch.mockClear();
    });

    it('attempts to fetch an access token if access token is not yet fetched (undefined)', async () => {
      await api.get('http://test.com').fetch();
      expect(mockedFetch.mock.calls.length).toStrictEqual(2);
      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });

    it('does not fetch an access token if the token is defined and not expired', async () => {
      Object.defineProperty(api, 'accessToken', {
        value: 'token',
      });
      Object.defineProperty(api, 'tokenExpiry', {
        value: Date.now() + 1,
      });

      await api.get('http://test.com').fetch();
      expect(mockedFetch.mock.calls.length).toStrictEqual(1);
    });

    it.each`
      tokenExpiryGetter
      ${() => Date.now()}
      ${() => Date.now() - 1}
    `('attempts to fetch an access token if access token is expired', async ({ tokenExpiryGetter }) => {
      Object.defineProperty(api, 'accessToken', {
        value: 'token',
      });
      Object.defineProperty(api, 'tokenExpiry', {
        value: tokenExpiryGetter(),
      });

      await api.get('http://test.com').fetch();
      expect(mockedFetch.mock.calls.length).toStrictEqual(2);
    });
  });
});
