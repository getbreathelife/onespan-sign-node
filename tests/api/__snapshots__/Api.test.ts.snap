// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api Authorization header attempts to fetch an access token if access token is not yet fetched (undefined) 1`] = `
Array [
  "http://demo.com/apitoken/clientApp/accessToken",
  Object {
    "body": "{\\"type\\":\\"OWNER\\",\\"clientId\\":\\"clientId\\",\\"secret\\":\\"secret\\"}",
    "headers": Object {
      "accept": "application/json",
      "content-type": "application/json",
    },
    "method": "POST",
  },
]
`;
